### ISSUE TITLE
Improve application exit behavior

### ISSUE THREAD

**Issue by bdionne** [NCIEVS/nci-protege5#562]
Protégé prompts about saving the ontology on quitting, which seems to be part of the semantics of opening ontologies from a file.

On the other hand when an ontology is opened from the client/server it does nothing on exit. We've had instances where modelers lost classes, possibly because they had neglected to save edits.

This enhancement is to check for unsaved edits in the session history and prompt the user before exiting, allowing them to do a save.

**Comment by bdionne** [comment]
The protege main Application object installs a hook that is called on exit. This hook asks the ontology if it is `dirty` and if so puts up the warning you see on exit about saving.

The client/server seems to have no impact on this, and the oncology still is marked `dirty` when you perform edits that are pushed to the server. It never resets that bit.

I think the solution will be to see if we can determine the connection status on exit. If the client/server is in use then we should bypass this check and see if the Session History is non-empty. If so we can popup a similar warning that allows the user to cancel and go back and saver.

This will be a change on protégé core.

**Comment by bdionne** [comment]
Fixed in [tag 4.3.0-qa1](https://github.com/NCIEVS/protege/tree/4.3.0-qa1)

**Comment by bdionne** [comment]
This fails to work when the app is closed using the icon in the corner, though it works fine using the file/exit menu. A new version of `EditTab`, [nci-edit-tab-1.4.7](https://github.com/NCIEVS/nci-edit-tab/releases/tag/nci-edit-tab-1.4.7)

fixes this.

Thanks to [Gilberto](https://github.com/fragosog) for triaging and reporting this.

**Comment by singhik** [comment]
@bdionne  : Hi Bob, please push that Edit Tab to NCIEVS repository plugin registry , currently all repo points to that for getting updated Plugins . 
Also, is terminating the App via  CTRL+C  option in scope for this Warning message ?

**Comment by bdionne** [comment]
> Hi Bob, please push that Edit Tab to NCIEVS repository plugin registry

I did @singhik, that's what the [link above ^^](https://github.com/NCIEVS/nci-edit-tab/releases/tag/nci-edit-tab-1.4.7) is. Just point your autoupdater to my registry and you'll receive it.

> Also, is terminating the App via CTRL+C option in scope for this Warning message ?

I doubt this would work, it's outside the application. Protege might be catching signals from the OS. It's easy enough to test.

**Comment by singhik** [comment]
@bdionne  : The Plugin registry URL points to NCIEVS repo and not bdionne repo , i was able to get the 1.4.7 by pointing to bdionee but the Auto Updater never points to bdionee but NCIEVS, cause bdionne always has +1 versions due to ongoing development .

I tested CTRL+C before posting above comment and that is called 'Actual Behaviour', i asked the question in the Ticket so that i can get the Requirement i.e 'Expected Behaviour' . Currently 'Actual Behaviour' does not prompt for any Save Ontology Warning Message .

**Comment by bdionne** [comment]
> The Plugin registry URL points to NCIEVS repo and not bdionne repo , i was able to get the 1.4.7 by pointing to bdionee but the Auto Updater never points to bdionee but NCIEVS, cause bdionne always has +1 versions due to ongoing development .

Correct, this is the procedure we all agreed to ages ago. I will push the new version of the plugins to the top level in the repos and they will then be *promoted* as required to QA, Stage, and Prod. We can review this at weekly status if you'd like to use some other process.

With respect to Ctl-C I'll leave that up to Gilberto. I wrote this ticket up, based on email from Liz, and my understanding was that they wanted to catch the situation of unsaved edits when they exit, which normally means click the icon in the upper right corner or use the File/Exit menu items.

We could probably add hooks as well to catch signals from the OS, but I'm not sure this is necessary. My sense is that when the user reaches the point of hitting Ctl-C it's because the app is wedged and already in an unrecoverable state.

**Comment by fragosog** [comment]
Hi @singhik , exiting with ^C would be out of scope for the exit message.  

I forgot to send you an email earlier this week.  It's probably best for now to get the jar directly from @bdionne 's [repo](https://github.com/NCIEVS/nci-edit-tab/releases/tag/nci-edit-tab-1.4.7), just drop the jar into the plugins directory after moving the older jar out of the way.

@chaozhang-nci , when you have a chance please create a new distribution with the updated edit-tab 1.4.7 jar.

**Comment by singhik** [comment]
This is tested on 4.3.0-qa1 build and is working as expected

**Comment by singhik** [comment]
This implementation has been Tested on 4.3.0-qa2 Build and is working as Expected for all Scenarios .

**Comment by bdionne** [comment]
In UAT another edge case was discovered, using the server menu to logout before the app exit. 

This will be addressed in the 4.4.0 milestone. With the metaproject admin plugin being separated out into a new app, the server login/logout menu items will go away. Users will only login using file/open from server

**Comment by kondareddys2** [comment]
The issue has been sucessfully verified in QA for Protege 4.4.0 milestone.

**Comment by bdionne** [comment]
Not really a bug, but the user wants an additional case handled, when `Server/Logout` is used to exit the project. This doesn't really exit protege but it does close the project so uncommitted edits should be warned about.

**Comment by bdionne** [comment]
fixed in [4.4.0-dev8](https://github.com/NCIEVS/protege/tree/4.4.0-dev8)
